enhancements/dns/dns-operator-operand-logging-level.md

```diff
@@ -0,0 +1,230 @@
+error: SERVFAIL, NOTIMP, REFUSED, etc. Anything that indicates the remote server is not willing to resolve the request.
+all: the default - nothing is specified. Using of this class means that all messages will be logged whatever we mix together with "all".
+
+Also, a logging level API for the DNS Operator would assist OpenShift developers working on the DNS Operator who may
+desire more in-depth logging statements when working on the operator's controllers.
+
+Additionally, a logging level API for CoreDNS logs would assist cluster administrators who wish to have more control
+over the output of their DNSController's CoreDNS logs.
+
+### Goals
+
+* Add a user-facing API for controlling the run-time verbosity of the [OpenShift DNS Operator and CoreDNS](https://github.com/openshift/cluster-dns-operator)
+
+### Non-Goals
+
+* Change the default logging verbosity of the DNS Operator or the CoreDNS in production OCP clusters.
+
+## Proposal
+
+### DNS Operator Log Level API
+We will be creating an API for **field** `operatorLogLevel` in DNSSpec in accordance with [LogLevel](https://github.com/openshift/api/blob/master/operator/v1/types.go#L71) type.
+Valid values will be : "Normal", "Debug", "Trace", "TraceAll".
+We will using [logrus](https://github.com/sirupsen/logrus#level-logging) to enable the log levels for the operator level logging.
+Logrus has seven logging levels: Trace, Debug, Info, Warning, Error, Fatal and Panic.
+log.Trace("Something very low level.")
+log.Debug("Useful debugging information.")
+log.Info("Something noteworthy happened!")
+log.Warn("You should probably take a look at this.")
+log.Error("Something failed but I'm not quitting.") // Calls os.Exit(1) after logging log.Fatal("Bye.") // Calls panic() after logging log.Panic("I'm bailing.")
+After the logging level on a Logger is set, log entries with that severity or anything above it will be logged.
+Eg- Default log.SetLevel(log.InfoLevel) will log anything that is info or above (warn, error, fatal, panic).
+So, we will be setting operatorLogLevel in a separate controller to watch dnses and set log level.
+operatorLogLevel: "Normal" will set logrus.SetLogLevel("Info")
+operatorLogLevel: "Debug" will set logrus.SetLogLevel("Debug")
+operatorLogLevel: "Trace" will set  logrus.SetLogLevel("Trace")
+
+
+### CoreDNS Log Level API
+
+Valid values for logLevel are: "Normal", "Debug", "Trace", "TraceAll" as per [Operator Api](https://github.com/openshift/api/blob/master/operator/v1/types.go#L62)
+
+We will be enabling the CLASSES(https://github.com/coredns/coredns/tree/master/plugin/log#syntax) of coredns w.r.t to the logLevel we have defined in openshift api.
+
+So,
+
+logLevel "Normal"  will enable errors class as we enable errors plugin by default log . { class error }
+
+logLevel "Debug" will enable log . { class denial error }
+
+logLevel "Trace" will enable  log . { class all }
```

Comment:
What is the relationship between the `errors` plugin and the `log . { class error }` stanza?  I think we discussed this before—doesn't `errors` log things like connection errors whereas `log . { class error }` logs error responses?
```suggestion
`logLevel: "Normal"`  will enable the "errors" class: `log . { class error }`.

`logLevel: "Debug"` will enable the "denial" class: `log . { class denial error }`.

`logLevel: "Trace"` will enable the "all" class: `log . { class all }`.

Note that the `errors` plugin is always enabled.  The `errors` plugin logs TCP/UDP connection errors whereas `log . { class error }` logs DNS error responses (such as SERVFAIL).  
```

This would be a good place to add a note that CoreDNS reloads its configuration without requiring a restart, so the operator can adjust CoreDNS's log level just by updating the Corefile configmap without need to restart the pod.